### PR TITLE
content-creator updates

### PR DESCRIPTION
**Pull request by roysagi**
## Status
- [x] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
related: https://github.com/demisto/etc/issues/24383
fixes: https://github.com/demisto/etc/issues/25276

## Description

* content creator now adds fromversion to all yml and json files.
* added parameter suffix (-s) to add suffix to all created files.
* improved outputs for create content to deliver a better status report.
* content_new zip now does not contain files (both yml and json) with 6.0.0 fromversion.

### CODE REVIEW

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 1]
```diff
@@ -98,6 +105,46 @@ def run(self):
 
         return 0
 
+    def add_from_version_to_yml(self, file_path, yml_content=None, save_yml=True):
+        if not yml_content:
+            yml_content = get_yaml(file_path)
+
+        if 'fromversion' not in yml_content or \
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 2]
```diff
@@ -98,6 +105,46 @@ def run(self):
...
+            if save_yml:
+                with open(file_path, 'w') as f:
+                    yaml.dump(yml_content, f)
+
+        return yml_content
+
+    def add_from_version_to_json(self, file_path):
+        json_content = None
+        with open(file_path, 'r') as f:
+            json_content = json.loads(f.read())
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 3]
```diff
@@ -98,6 +105,46 @@ def run(self):
...
+
+        if json_content is not None and \
+                ('fromVersion' not in json_content or
+                 parse_version(self.fromversion) > parse_version(json_content.get('fromVersion', '0'))):
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: solved by the change @yaakovi suggested

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 4]
```diff
@@ -191,7 +239,10 @@ def copy_dir_yml(self, dir_path, bundle):
         scan_files, _ = get_yml_paths_in_dir(dir_path, error_msg='')
         content_files = 0
         dir_name = os.path.basename(dir_path)
+        if len(scan_files) > 0:
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 5]
```diff
@@ -222,37 +282,46 @@ def copy_dir_json(self, dir_path, bundle):
...
-                    f'Failed while trying to create {os.path.join(bundle, new_path)}. File already exists.'
-                )
-            dpath = new_path
-
-            if len(dpath) >= self.file_name_max_size:
-                self.long_file_names.append(os.path.basename(dpath))
-
-            shutil.copyfile(path, os.path.join(bundle, dpath))
+        count_files = 0
+        if len(scan_files) > 0:
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:276]: but if there are no files to scan this creates a lot of unnecessary noise in the output don't you think?

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 6]
```diff
@@ -222,37 +282,46 @@ def copy_dir_json(self, dir_path, bundle):
...
+                        dpath = f'incidentfield-{dpath}'
+                if dir_name == 'Dashboards':
+                    if not dpath.startswith('dashboard-'):
+                        dpath = f'dashboard-{dpath}'
+                if dir_name == 'Layouts':
+                    if not dpath.startswith('layout-'):
+                        dpath = f'layout-{dpath}'
+                new_path = dpath
+                if dir_name == 'IndicatorFields' and not dpath.startswith('incidentfield-indicatorfield-'):
+                    new_path = dpath.replace('incidentfield-', 'incidentfield-indicatorfield-')
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:299]: indeed - no changes made here

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 7]
```diff
@@ -265,18 +334,24 @@ def copy_dir_md(self, dir_path, bundle):
...
-                    raise NameError(
-                        f'Failed while trying to create {os.path.join(bundle, new_path)}. File already exists.'
-                    )
-
-            if len(new_path) >= self.file_name_max_size:
-                self.long_file_names.append(os.path.basename(new_path))
-
-            shutil.copyfile(path, os.path.join(bundle, new_path))
+        count_files = 0
+        if len(scan_files) > 0:
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:328]: same question as above

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 8]
```diff
@@ -531,7 +615,8 @@ def create_content(self, only_packs=False):
             self.copy_file_to_artifacts('release-notes.md')
             self.copy_file_to_artifacts('beta-release-notes.md')
             self.copy_file_to_artifacts('packs-release-notes.md')
-            print(f'finished creating the content artifacts at "{os.path.abspath(self.artifacts_path)}"')
+            print_color(f'\nfinished creating the content artifacts at "{os.path.abspath(self.artifacts_path)}"',
+                        LOG_COLORS.GREEN)
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 9]
```diff
@@ -133,7 +180,8 @@ def create_unifieds_and_copy(self, package_dir, dest_dir='', skip_dest_dir=''):
                 # for example HelloWorld integration
                 unification_tool = Unifier(package, package_dir_name, skip_dest_dir)
                 print('skipping {}'.format(package))
-            unification_tool.merge_script_package_to_yml()
+            unified_yml_path = unification_tool.merge_script_package_to_yml()[0]
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: point taken - I'll change it to check both

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 10]
```diff
@@ -200,16 +251,25 @@ def copy_dir_yml(self, dir_path, bundle):
             with io.open(path, mode='r', encoding='utf-8') as file_:
                 yml_info = ryaml.load(file_)
             ver = yml_info.get('fromversion', '0')
-            print(f' - processing: {ver} ({path})')
+            updated_yml_info = self.add_from_version_to_yml(file_path=None, yml_content=yml_info, save_yml=False)
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: I'll add a default value of None

File: demisto_sdk/commands/create_artifacts/tests/content_creator_test.py [hunk 11]
```diff
@@ -185,3 +194,135 @@ def test_copy_file_to_artifacts(self):
         file_path = os.path.join(self._files_to_artifacts_dir, filename)
         content_creator.copy_file_to_artifacts(file_path)
         assert filecmp.cmp(file_path, os.path.join(self.content_repo, filename))
+
+    def test_add_from_version_to_yml__no_fromversion_in_yml(self, repo):
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/tests/content_creator_test.py:198]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 12]
```diff
@@ -9,7 +9,10 @@
 import zipfile
 from typing import List
 
+from pkg_resources import parse_version
```
[the remark this answers is not in the document]
Comment (yaakovi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:12]: no need, just curious

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:276]: you can remove the prints as well then.

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:328]: same response as above ;)

**Comment by roysagi** [comment]
> > if you add `fromversion` automatically to all content items with (currently) 4.5 so the resulting zip will contain no content for 4.1.
> > not sure this is the desired behavior
> 
> nice work, please respond to my head comment from the last review

The idea is to merge this while moving 4.1 to a dedicated branch. Hence moving forward the created content zips should hold only integrations from version 4.5 - this is where this change comes to use.

So the lack of 4.1 is intentional.

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 13]
```diff
@@ -148,17 +224,17 @@ def add_tools_to_bundle(tools_dir_path, bundle):
...
+    def copy_playbook_yml(self, path, out_path):
         """
         Add "playbook-" prefix to playbook file's copy destination filename if it wasn't already present
         """
         dest_dir_path = os.path.dirname(out_path)
         dest_file_name = os.path.basename(out_path)
         if not dest_file_name.startswith('playbook-') and tools.find_type(path) == 'playbook':
             new_name = '{}{}'.format('playbook-', dest_file_name)
-            out_path = os.path.join(dest_dir_path, new_name)
+            out_path = self.add_suffix_to_file_path(os.path.join(dest_dir_path, new_name))
```
Comment (ronykoz) [inline demisto_sdk/commands/create_artifacts/content_creator.py:235]: why do we need the suffix

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 14]
```diff
@@ -191,25 +267,39 @@ def copy_dir_yml(self, dir_path, bundle):
...
-                # in TestPlaybook dir we might have scripts - all should go to test_bundle
-                if dir_name == 'TestPlaybooks' and os.path.basename(path).startswith('script-'):
-                    self.copy_content_yml(path, os.path.join(bundle, os.path.basename(path)), yml_info)
-                self.copy_playbook_yml(path, os.path.join(bundle, os.path.basename(path)))
-            else:
-                self.copy_content_yml(path, os.path.join(bundle, os.path.basename(path)), yml_info)
-            content_files += 1
-        print(f' - total files: {content_files}')
+            if self.check_from_version(path):
+                new_file_path = self.add_suffix_to_file_path(os.path.join(bundle, os.path.basename(path)))
```
Comment (ronykoz) [inline demisto_sdk/commands/create_artifacts/content_creator.py:274]: same here

File: demisto_sdk/commands/unify/unifier.py [hunk 15]
```diff
@@ -137,13 +139,20 @@ def write_yaml_with_docker(self, yml_unified, yml_data, script_obj):
 
         return output_map
 
-    def merge_script_package_to_yml(self):
+    def merge_script_package_to_yml(self, file_name_suffix=None, filter_6_0=False):
         """Merge the various components to create an output yml file
         """
+        if filter_6_0 and parse_version(self.yml_data.get('fromversion', '0.0.0')) >= parse_version('6.0.0'):
```
Comment (ronykoz) [inline demisto_sdk/commands/unify/unifier.py:145]: why do we need to check this here and not before the merge too?

**Comment by ronykoz** [comment]
@yaakovi could you go over this as well please?

File: demisto_sdk/__main__.py [hunk 16]
```diff
@@ -299,6 +299,11 @@ def validate(config, **kwargs):
 @click.option(
     '--no-update-commonserver', is_flag=True, help='Whether to update CommonServerPython or not - used for local runs.'
 )
+@click.option(
+    '-s', '--suffix', help='The added suffix to all files in the created artifacts')
```
[the remark this answers is not in the document]
Comment (roysagi) [reply demisto_sdk/__main__.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 13 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:235]: as discussed we wanted to add the option to add a suffix to the file path in order to distinguish between the files once we merge 2 zip files (one from regular release branch and the other from 4.1 branch)

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 14 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py:274]: same answer as above

File: demisto_sdk/commands/unify/unifier.py (hunk 15 above)
Comment (roysagi) [reply demisto_sdk/commands/unify/unifier.py]: will move this to creator

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 17]
```diff
@@ -133,6 +139,37 @@ def add_from_version_to_json(file_path):
 
         return json_content
 
+    def add_suffix_to_file_path(self, file_path):
+        return os.path.splitext(file_path)[0] + self.file_name_suffix + os.path.splitext(file_path)[1]
+
+    def check_from_version(self, file_path):
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_creator.py:145]: function name doesn't explain its purpose very well...
check for what?

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 18]
```diff
@@ -133,6 +139,37 @@ def add_from_version_to_json(file_path):
...
+            if parse_version(yml_content.get('fromversion', '0.0.0')) >= parse_version('6.0.0'):
+                return False
+
+        elif file_path.endswith('.json'):
+            json_content = tools.get_json(file_path)
+            if parse_version(json_content.get('fromVersion', '0.0.0')) >= parse_version('6.0.0'):
+                return False
+
+        elif file_path.endswith('.md'):
+            self.check_md_related_from_version(file_path)
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_creator.py:157]:
```suggestion
            return self.check_md_related_from_version(file_path)
```

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 19]
```diff
@@ -233,32 +271,34 @@ def copy_dir_yml(self, dir_path, bundle):
...
-                yml_info = updated_yml_info
-
-            process_message = f' - processing: {path}'
-            if ver != '0' and ver != '':
-                process_message += f' - current fromversion: {ver}'
-            print(process_message)
-            if dir_name in ['Playbooks', 'TestPlaybooks']:
-                # in TestPlaybook dir we might have scripts - all should go to test_bundle
-                if dir_name == 'TestPlaybooks' and os.path.basename(path).startswith('script-'):
+            if self.check_from_version(path):
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_creator.py:273]: will minimize indentantion levels (same for all uses).
```suggestion
            if not self.check_from_version(path):
                continue
```

File: demisto_sdk/commands/create_artifacts/content_creator.py [hunk 20]
```diff
@@ -376,28 +419,31 @@ def copy_test_files(self, test_playbooks_dir=TEST_PLAYBOOKS_DIR):
             if os.path.isdir(path):
                 non_circle_tests = glob.glob(os.path.join(path, '*'))
                 for new_path in non_circle_tests:
-                    print(f'copying path {new_path}')
-                    new_file_path = os.path.join(self.test_bundle, os.path.basename(new_path))
-                    shutil.copyfile(new_path, new_file_path)
-                    self.add_from_version_to_yml(new_file_path)
+                    if os.path.isfile(new_path) and self.check_from_version(new_path):
```
Comment (yaakovi) [inline demisto_sdk/commands/create_artifacts/content_creator.py:421]: when will glob return a file that doesn't exist?
or is that for skipping directories?

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 17 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 18 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 20 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: skipping directories from what I gather - I did not change that part

File: demisto_sdk/commands/unify/unifier.py [hunk 21]
```diff
@@ -143,13 +143,17 @@ def write_yaml_with_docker(self, yml_unified, yml_data, script_obj):
...
+    def merge_script_package_to_yml(self, file_name_suffix=None):
         """Merge the various components to create an output yml file
         """
         print("Merging package: {}".format(self.package_path))
         package_dir_name = os.path.basename(self.package_path)
         output_filename = '{}-{}.yml'.format(DIR_TO_PREFIX[self.dir_name], package_dir_name)
 
+        if file_name_suffix:
+            output_filename = ''.join(output_filename.split('.')[:-1]) + file_name_suffix + \
+                              '.' + output_filename.split('.')[-1]
```
Comment (yaakovi) [inline demisto_sdk/commands/unify/unifier.py:155]:
```suggestion
            # append suffix to output file name
            output_filename = file_name_suffix.join(os.path.splitext(output_filename))
```

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 19 above)
Comment (yaakovi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: @roysagi you ignored this comment

File: demisto_sdk/commands/unify/unifier.py (hunk 21 above)
Comment (roysagi) [reply demisto_sdk/commands/unify/unifier.py:155]: done

File: demisto_sdk/commands/create_artifacts/content_creator.py (hunk 19 above)
Comment (roysagi) [reply demisto_sdk/commands/create_artifacts/content_creator.py]: missed it. done now

**Comment by roysagi** [comment]
@ronykoz @yaakovi - Maybe we should merge with the "Latest_support_version" set to 4.1.0 until we create the feature branch for that version - what do you think?

**Comment by yaakovi** [comment]
> @ronykoz @yaakovi - Maybe we should merge with the "Latest_support_version" set to 4.1.0 until we create the feature branch for that version - what do you think?

agreed.
as long as the content zip merging doesn't work we. can't remove the 4.1 files from the zip.

**Pull request merged**